src/coinbase/address/wallet_address.ts

```diff
@@ -322,22 +337,266 @@ export class WalletAddress extends Address {
    */
-  private async validateCanTrade(amount: Amount, fromAssetId: string) {
+  private async validateAmount(amount: Amount, assetId: string) {
     if (!this.canSign()) {
       throw new Error("Cannot trade from address without private key loaded");
     }
-    const currentBalance = await this.getBalance(fromAssetId);
+    const currentBalance = await this.getBalance(assetId);
     amount = new Decimal(amount.toString());
     if (currentBalance.lessThan(amount)) {
       throw new Error(
         `Insufficient funds: ${amount} requested, but only ${currentBalance} available`,
       );
     }
   }
+
+  /**
+   * Creates a staking operation to stake, signs it, and broadcasts it on the blockchain.
+   *
+   * @param amount - The amount for the staking operation.
+   * @param assetId - The asset to the staking operation.
+   * @param action - The type of staking action to perform.
+   * @param timeoutSeconds - The amount to wait for the transaction to complete when broadcasted.
+   * @param intervalSeconds - The amount to check each time for a successful broadcast.
+   * @param options - Additional options such as setting the mode for the staking action.
+   * @returns The staking operation after it's completed fully.
+   */
+  public async createStakingOperation(
+    amount: Amount,
+    assetId: string,
+    action: string,
+    timeoutSeconds = 60,
+    intervalSeconds = 0.2,
+    options: CoinbaseWalletAddressStakeOptions = { mode: StakeOptionsMode.DEFAULT },
+  ): Promise<StakingOperation> {
+    await this.validateAmount(amount, assetId);
+    let stakingOperation = await this.createStakingOperationRequest(
+      amount,
+      assetId,
+      action,
+      options,
+    );
+
+    // NOTE: Staking does not yet support server signers at this point.
+    await stakingOperation.sign(this.key!);
+    for (const tx of stakingOperation.getTransactions()) {
+      if (!tx.isSigned()) {
+        continue;
+      }
+      stakingOperation = await this.broadcastStakingOperationRequest(
+        stakingOperation,
+        tx.getSignedPayload()!.slice(2),
+      );
+    }
+
+    const startTime = Date.now();
+    while (Date.now() - startTime < timeoutSeconds * 1000) {
+      await stakingOperation.reload();
+      const status = stakingOperation.getStatus();
+      if (status === StakingOperationStatus.COMPLETE || status === StakingOperationStatus.FAILED) {
+        return stakingOperation;
+      }
+      await delay(intervalSeconds);
+    }
+    throw new Error("Staking Operation timed out");
+  }
+
+  /**
+   * Get the stakeable balance for the supplied asset.
+   *
+   * @param asset_id - The asset to check the stakeable balance for.
+   * @param mode - The staking mode. Defaults to DEFAULT.
+   * @param options - Additional options for getting the stakeable balance.
+   * @returns The stakeable balance.
+   */
+  public async stakeableBalance(
```

Comment:
can we move these calls common b/w the 2 models to the parent class address.ts ?

Reply:
+1

Reply:
Good point, done!